### ISSUE TITLE
SCSS variables order

### ISSUE THREAD

**Issue by borodean** [stylelint/stylelint#518]
Since $variables are getting interpreted as properties by [postcss-scss](https://github.com/postcss/postcss-scss), stylelint takes them into account when checking for [`rule-properties-order`](http://stylelint.io/?/src/rules/rule-properties-order/README.md) rule. If you set it to `alphabetical`, stylelint complains about variables not being sorted, which makes no sense if their order is determined by the order of computations:

```scss
@mixin size($size) {
  // You don't want these variables to be automatically sorted:
  $size: unpack($size, 2);
  $important: if(length($size) == 3, !important, null);
  $width: nth($size, 1);
  $height: nth($size, 2);
  width: if($width != none, $width $important, null);
  height: if($height != none, $height $important, null);
} 
```

However, when variables are not involved in any algorythms, you *may* want them to be sorted:

```scss
.masthead {
  // These variables order doesn't matter,
  // could be automatically sorted:
  $background: #c00;
  $height: 120px;
  background: $background;
  height: $height;
  line-height: $height;

  &_inner {
    background: invert($background);
    color: $background;
  }
}
```

What do you think could be done about that?

cc @ai @ben-eb

**Comment by jeddy3** [comment]
@davidtheclark I'm not sure what you think, but we've discussed non-standard syntax before and we concluded that "however, since there are infinite processing possibilities, the linter cannot support everything." (ref [css processor doc](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md)). So, I suggest keeping the `rule-properties-order` rule the same and if someone wants to build this functionality, then they can do so in a plugin, perhaps called `rule-scss-properties-order`. As I can see the scope of that rule ballooning as more scss syntax gets requested. A user can then choose either to not use scss vars inline and use the `rule-properties-order` rule, or use scss vars inline and not lint it or write a plugin to support this syntax.

What do you think?

**Comment by davidtheclark** [comment]
Definitely if it the user wants to sort SCSS variables that should be an external plugin.

We could modify `rule-properties-order` so that at least it doesn't complain. We could make it so that by default it ignores properties that start with `$` or `@`; or we could make that an option, `ignore-sass-less-variables`.

As I was thinking in https://github.com/stylelint/stylelint/issues/438: There are indeed infinite processing possibilities, but I think that simple accommodations to `$` and `@` variable syntax here and there are legit, as they will help out so many possible users.

**Comment by jeddy3** [comment]
> There are indeed infinite processing possibilities, but I think that simple accommodations to $ and @ variable syntax here and there are legit

Yep, you're right. That is how we've been approaching it.

> We could modify rule-properties-order so that at least it doesn't complain. We could make it so that by default it ignores properties that start with $ or @;

Ignoring by default sgtm as they're more _variables_ than _properties_, and the rule is called `rule-`_`properties`_`-order` :)

@borodean Thanks for reporting and for the suggestion. Feel free to create a plugin to handle scss variable sorting. 

@davidtheclark Shall we close this issue once `$` and `@` are ignored?

**Comment by davidtheclark** [comment]
> Shall we close this issue once $ and @ are ignored?

Yep, that's what I'd think.

**Comment by davidtheclark** [comment]
Fixed in c9d1967
